### PR TITLE
Further container logic [RDY]

### PR DESCRIPTION
**Pull request by mugling**
**EDIT: rebased against master**

~~This PR continues directly from #14949 which is not yet merged. There are 11 new commits starting with 09ff0fd~~. It contains extensions to `item_location` and `*::visit_items` sufficient to implement this:

ccda

**This PR does not implement a UI for nested containers** but does however provide some of the key functionality that would be required. `Character::find_parent` can be used to determine whether an item has a parent container from which it can then be recursively unpacked via `item_location::obtain`.

The above screenshot demonstrates this functionality:
- `STANAG magazine (30)` is contained by `drop leg ammo pouch` and worn by `Character`  
- `item_location::describe` uses `Character::find_parent` to help determine the container name
- `game::reload` uses `item_location::obtain` to unpack the item to the base of the inventory
- Reloading then proceeds as normal via `player::assign_activity`

Implementation details:
- `*::visit_items` is extended to also provide a pointer to the parent node
- `*::find_parent` uses the above to find any parent container
- `item_location::describe` can optionally return the location relative to the player
- `item_location::obtain` fetches the item to the root of the character inventor and in doing so it triggers events, eg. `item::on_takeoff` and deducts the appropriate number of moves

**tl;dr**
Demonstrates a wearable ammo pouch from which you can reload

### CODE REVIEW

File: data/json/items/magazines.json [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        "volume" : 0,
+        "weight" : 0,
+        "to_hit" : -4,
+        "rigid": false,
+        "ammo_type" : "223",
+        "capacity" : 200,
+        "count" : 100,
+        "reliability" : 8,
+        "reload_time" : 80,
+        "flags": [ "NO_RELOAD", "NO_UNLOAD" ]
```
Comment (Coolthulhu) [inline data/json/items/magazines.json]: Why no unload? Can't you get rid of the belt and just use the ammo without it?

Comment (mugling) [inline data/json/items/magazines.json]: Too easy to accidentally destroy your ammo belt. I was going to allow disassembly though.

File: src/character.cpp [hunk 2]
```diff
@@ -484,26 +484,46 @@ VisitResponse Character::visit_items( const std::function<VisitResponse( item& )
     return inv.visit_items( func );
 }
 
-VisitResponse Character::visit_items( const std::function<VisitResponse( const item& )>& func ) const
+VisitResponse Character::visit_items( const std::function<VisitResponse( const item *, const item * )>& func ) const
```
Comment (Coolthulhu) [inline src/character.cpp]: Looks like you're overly specifying the visit function for that one usage.
In most invocations the second argument will be unused. There are 64 matches for "visit_items" on this page, but only one of them actually makes use of the second argument.

Can't it be done in some other way? At the very least by adding the two argument function as an overload rather than making it the only way to do it.

File: src/character.cpp [hunk 3]
```diff
@@ -484,26 +484,46 @@ VisitResponse Character::visit_items( const std::function<VisitResponse( item& )
...
+    if( visit_items( [&]( item *node, item *parent ){
+        if( node == &it ) {
+            res = parent;
+            return VisitResponse::ABORT;
+        }
+        return VisitResponse::NEXT;
+    } ) != VisitResponse::ABORT ) {
+        debugmsg( "Tried to find item parent using character who doesn't have the item" );
+    }
+    return res;
```
Comment (Coolthulhu) [inline src/character.cpp]: It would be a tiny bit safer if the return value of it was a reference instead. We already check for `item.is_null()` all over the code so it wouldn't be unusual to have this one return a null item (`ret_null`) rather than a pointer.

**Comment by Coolthulhu** [comment]
I don't like the idea of changing the drop leg pouches purely to ammo pouches.
It is an useful storage item that doesn't have an alternative. Ammo pouches should be a different item.

**Comment by Rivet-the-Zombie** [comment]
>Ammo pouches should be a different item.

I'm going to second this.

File: src/character.cpp (hunk 2 above)
Comment (mugling) [inline src/character.cpp]: > 64 matches for "visit_items" on this page

Many of these are declarations as three classes implement their own version (plus a const overload) and then delegate to the next implementation. There's approximately a dozen actual calls to the function.

> Looks like you're overly specifying the visit function for that one usage.

`*::visit_items` is intended to provide the primitives (lookup, count, delete etc). As we add remainder of these eg. `*::remove_items_with` most will depend on this extra parameter

> adding the two argument function as an overload

This would work but I'm not sure if it's necessarily any cleaner as we now have four overloads of each function not two.

I agree however that the higher order functions should all use unary predicates eg. `std::function<bool(item&)>`.

File: src/character.cpp (hunk 3 above)
Comment (mugling) [inline src/character.cpp]: If we implement `inventory::find_parent` and `item::find_parent` where does `ret_null` come from?

**Comment by mugling** [comment]
> I don't like the idea of changing the drop leg pouches purely to ammo pouches.

I'll create a new type but before I do so are the JSON values about right?

File: src/character.cpp (hunk 3 above)
Comment (Coolthulhu) [inline src/character.cpp]: OK, that wouldn't work.
We probably should have a global null item some day, but we don't have one yet.

Comment (mugling) [inline src/character.cpp]: Global null item is an excellent idea and I can't see any major obstacles to implementing it?

File: data/json/items/armor.json [hunk 4]
```diff
@@ -10594,26 +10594,34 @@
...
-        "name" : "pair of drop leg pouches",
-        "name_plural" : "pairs of drop leg pouches",
-        "weight" : 205,
+        "name" : "drop leg ammo pouch",
+        "name_plural" : "drop leg ammo pouches",
+        "weight" : 100,
         "color" : "dark_gray",
-        "covers" : ["LEGS"],
+        "covers" : ["LEG_EITHER"],
         "to_hit" : 0,
```
Comment (Coolthulhu) [inline data/json/items/armor.json]: "to_hit" is optional

File: data/json/items/armor.json [hunk 5]
```diff
@@ -10594,26 +10594,34 @@
...
+        "storage" : 0,
         "symbol" : "[",
-        "description" : "A set of pouches that can be worn on the thighs using buckled straps.  This variety is favored by the military.",
-        "price" : 3000,
+        "description" : "An ammo pouch that can be worn on the thigh using buckled straps.  This variety is favored by the military.",
+        "price" : 2000,
         "material" : ["cotton", "null"],
-        "volume" : 2,
+        "volume" : 1,
         "cutting" : 0,
```
Comment (Coolthulhu) [inline data/json/items/armor.json]: "cutting" is optional

File: data/json/items/armor.json [hunk 6]
```diff
@@ -10594,26 +10594,34 @@
...
         "warmth" : 0,
         "phase" : "solid",
-        "encumbrance" : 5,
+        "encumbrance" : 2,
         "bashing" : -3,
```
Comment (Coolthulhu) [inline data/json/items/armor.json]: "bashing" below 0 doesn't really make sense and so can be removed too

File: src/character.cpp (hunk 3 above)
Comment (Coolthulhu) [inline src/character.cpp]: I thought that it might need something loaded. I checked now and it looks safe.
Could be declared in item.h, either as extern or a static method that returns a reference to the static null item.

**Comment by Rivet-the-Zombie** [comment]
>I'll create a new type but before I do so are the JSON values about right?

The current values look good for such an application.

**Comment by mugling** [comment]
Better name than *drop leg ammo pouch*?

**Comment by Rivet-the-Zombie** [comment]
How about 'leg ammo pouch' so we could then add other variants like 'chest ammo carrier'?

**Comment by mugling** [comment]
Rebased against master as there was a merge conflict with #14949

File: data/json/items/armor.json (hunk 6 above)
Comment (illi-kun) [inline data/json/items/armor.json]: @Coolthulhu why it doesn't make sense? I suppose it's harder to smash something by item with ```bashing = -3``` than one with ```bashing = 0``` (if I got it right from code) or you mean it doesn't make sense from common sense point of view?

File: data/json/items/armor.json [hunk 7]
```diff
@@ -10619,6 +10619,35 @@
...
+        "color" : "dark_gray",
+        "covers" : ["LEG_EITHER"],
+        "storage" : 0,
+        "symbol" : "[",
+        "description" : "A small fabric ammo pouch that can be strapped to your leg and capable of holding a single magazine close at hand.",
+        "price" : 2000,
+        "material" : ["cotton", "null"],
+        "volume" : 1,
+        "warmth" : 0,
+        "phase" : "solid",
```
Comment (illi-kun) [inline data/json/items/armor.json]: ```"phase" : "solid",``` is default value for optional entry

File: data/json/items/armor.json [hunk 8]
```diff
@@ -10619,6 +10619,35 @@
...
+        "encumbrance" : 3,
+        "flags" : ["VARSIZE", "WATER_FRIENDLY", "BELTED"],
+        "coverage" : 15,
+        "material_thickness" : 2,
+        "use_action" : {
+            "type": "holster",
+            "holster_prompt": "Stash ammo",
+            "holster_msg": "You stash your %s",
```
Comment (illi-kun) [inline data/json/items/armor.json]: Please put a period at the end.

File: data/json/items/armor.json [hunk 9]
```diff
@@ -10619,6 +10619,35 @@
...
+            "max_volume": 2,
+            "draw_cost": 5,
+            "flags": [ "COMPACT_MAG" ]
```
Comment (illi-kun) [inline data/json/items/armor.json]: Why this flag is here? Do we need to have the same flag on a container and on a magazine we want to store? If so, it's better to announce it in docs.

Comment (mugling) [inline data/json/items/armor.json]: > Do we need to have the same flag on a container and on a magazine we want to store?

Yes

> If so, it's better to announce it in docs.

See 4842355. Is there anywhere else?

File: src/character.cpp (hunk 3 above)
Comment (mugling) [inline src/character.cpp]: A good idea but separate PR

File: data/json/items/armor.json (hunk 9 above)
Comment (illi-kun) [inline data/json/items/armor.json]: I saw the referenced note but I misunderstood it. Anyway, thanks for clarification.

>  Is there anywhere else?

No, I'm talking about exactly this place.

File: data/json/items/armor.json (hunk 6 above)
Comment (Coolthulhu) [inline data/json/items/armor.json]: It is ignored anyway, because it doesn't make much sense. It only ever makes sense to attack shockers with items with stats this bad, for the rest fists are better.
In `item::damage_bash/cut`, only non-negative values are returned.

File: data/json/items/armor.json (hunk 9 above)
Comment (mugling) [inline data/json/items/armor.json]: No worries, it's all a bit jumbled after rebasing

**Comment by chaosvolt** [comment]
Ah, interesting.

File: src/item_location.cpp [hunk 10]
```diff
@@ -189,6 +299,34 @@ class item_location::item_on_vehicle : public item_location::impl {
...
+        if( !what ) {
+            return INT_MIN;
+        }
+
+        int mv = 0;
+
+        // @todo handle unpacking costs
+        // @todo account for distance
+
+        mv += dynamic_cast<player *>( &ch )->item_handling_cost( *what );
```
Comment (BevapDin) [inline src/item_location.cpp]: Btw. you can dynamic cast reference, no need to convert the value to a pointer first.

File: src/character.cpp (hunk 2 above)
Comment (BevapDin) [inline src/character.cpp]: Why is the first parameter a pointer now? Do you expect it be null at any time?

File: src/item_location.cpp (hunk 10 above)
Comment (mugling) [inline src/item_location.cpp]: Are you sure this works on all compilers?

Comment (BevapDin) [inline src/item_location.cpp]: I don't know, but The C++ Standard says so.

**Comment by mugling** [comment]
Found problem during testing, working on fix

**Comment by chaosvolt** [comment]
Oh? o.o

**Comment by mugling** [comment]
Fixed in 3504233
